### DISCUSSION TITLE
Potential bug when removing files

### DISCUSSION THREAD

**Discussion by sinnbeck** [filamentphp/filament#13757] [Help]
### Package

Form builder

### Package Version

v3.2.96

### How can we help you?

I have a test that suddenly fails after upgrading to the latest minor version. I think there might have been introduced a bug, but I havent been able to locate it.

I have a file upload that has the following code, meant to remove the file if it fails serverside validation.
```php
->afterStateUpdated(function (?array $state) {
      try {
          return $this->validateOnly('data.files');

      } catch (ValidationException $exception) {
          foreach ($state as $key => $file) {
              $this->removeFormUploadedFile('data.files', $key);
          }

          throw $exception;
      }
  })
```
Now the first issue is that it does not seem to remove the file anymore?

And secondly, my tests are failing, as it changes state to and array with null in it `[null]` instead of `[]`

In regard to the second, issue it strangely seems to clean it correctly here https://github.com/filamentphp/filament/blob/3.x/packages/forms/src/Components/BaseFileUpload.php#L655
But it then calls `afterStateUpdated()` again, where I can see that it has added in null to the array

This is the line that throws an error as it does not allow `null`
https://github.com/filamentphp/filament/blob/3.x/packages/forms/src/Components/BaseFileUpload.php#L596

**Comment by sinnbeck** [comment]
Seems to have been broken since 3.2.77 and onwards. A different bug was introduced in that version and the fix in 3.2.83 seems to have caused this new bug

**Comment by sinnbeck** [comment]
One strange thing. If I use xdebug to see what is happening, then the browser only calls `afterStateUpdated` once, while the test is calling it twice. I suspect that this might be part of the reason it fails as I dont see any reason for why it should be called twice.

It also happen if I just call it regulary like this. The validation is called twice
```php
->afterStateUpdated(function (?array $state) {
        return $this->validateOnly('data.files');
})

**Reply by leandrocfe** [reply]
Did you test another field?

**Reply by sinnbeck** [reply]
Yeah I tested on a TextInput, and it only gets called once as expected. I have made an issue with a repo that shows the issue :)

https://github.com/filamentphp/filament/issues/13776
